FWCore/Framework/src/EventSetupRecordIOVQueue.cc

```diff
@@ -0,0 +1,132 @@
+
+namespace edm {
+  namespace eventsetup {
+
+    EventSetupRecordIOVQueue::EventSetupRecordIOVQueue(unsigned int nConcurrentIOVs)
+        : nConcurrentIOVs_(nConcurrentIOVs), iovQueue_(nConcurrentIOVs), cacheIdentifier_(1) {
+      for (unsigned int i = 0; i < nConcurrentIOVs; ++i) {
+        isAvailable_.push_back(std::make_unique<std::atomic<bool>>(true));
+      }
+      waitForIOVsInFlight_ = edm::make_empty_waiting_task();
+      waitForIOVsInFlight_->increment_ref_count();
+      waitForIOVsInFlight_->increment_ref_count();
+    }
+
+    EventSetupRecordIOVQueue::~EventSetupRecordIOVQueue() {
+      { WaitingTaskHolder tmp{std::move(endIOVTaskHolder_)}; }
+      waitForIOVsInFlight_->decrement_ref_count();
+      waitForIOVsInFlight_->wait_for_all();
+    }
+
+    void EventSetupRecordIOVQueue::setNewIntervalForAnySubProcess() {
+      bool newIntervalForAnySubProcess = false;
+      for (auto const& recordProvider : recordProviders_) {
+        if (recordProvider->intervalStatus() == EventSetupRecordProvider::IntervalStatus::NewInterval) {
+          newIntervalForAnySubProcess = true;
+          break;
+        }
+      }
+      for (auto& recordProvider : recordProviders_) {
+        recordProvider->setNewIntervalForAnySubProcess(newIntervalForAnySubProcess);
+      }
+    }
+
+    void EventSetupRecordIOVQueue::checkForNewIOVs(WaitingTaskHolder const& taskToStartAfterIOVInit,
+                                                   WaitingTaskList& endIOVWaitingTasks,
+                                                   bool newEventSetupImpl) {
+      for (auto& recordProvider : recordProviders_) {
+        if (recordProvider->newIntervalForAnySubProcess()) {
+          startNewIOVAsync(taskToStartAfterIOVInit, endIOVWaitingTasks);
+          return;
+        }
+      }
+
+      for (auto& recordProvider : recordProviders_) {
+        if (recordProvider->intervalStatus() != EventSetupRecordProvider::IntervalStatus::Invalid) {
+          endIOVWaitingTasks.add(endIOVWaitingTask_);
+          break;
+        }
+      }
+
+      for (auto& recordProvider : recordProviders_) {
+        recordProvider->continueIOV(newEventSetupImpl);
+      }
+    }
+
+    void EventSetupRecordIOVQueue::startNewIOVAsync(WaitingTaskHolder const& taskToStartAfterIOVInit,
+                                                    WaitingTaskList& endIOVWaitingTasks) {
+      ++cacheIdentifier_;
+      {
+        // Let the old IOV end when all the lumis using it are done.
+        // otherwise we'll deadlock when there is only one thread.
+        WaitingTaskHolder tmp{std::move(endIOVTaskHolder_)};
+      }
+      auto taskHolder = std::make_shared<WaitingTaskHolder>(taskToStartAfterIOVInit);
+      auto startIOVForRecord =
+          [this, taskHolder, &endIOVWaitingTasks](edm::LimitedTaskQueue::Resumer iResumer) mutable {
+            try {
+              unsigned int iovIndex = 0;
+              for (; iovIndex < nConcurrentIOVs_; ++iovIndex) {
```

Comment:
Wouldn't it just be better just use the length of the container?
```
```suggestion
              for (; iovIndex < isAvailable_.size(); ++iovIndex) {
```

Reply:
Done.